cmd/snap/cmd_run.go

```diff
@@ -530,6 +531,62 @@ func migrateXauthority(info *snap.Info) (string, error) {
 	return targetPath, nil
 }
 
+func activateXdgDocumentPortal(info *snap.Info) error {
+	// Don't do anything for snaps that don't plug the desktop interface
+	//
+	// NOTE: This check is imperfect because we don't really know
+	// if the interface is connected or not but this is an
+	// acceptable compromise for not having to communicate with
+	// snapd in snap run. In a typical desktop session the
+	// document portal can be in use by many applications, not
+	// just by snaps, so this is at most, pre-emptively using some
+	// extra memory.
+	plugsDesktop := false
+	for _, plug := range info.Plugs {
+		if plug.Interface == "desktop" {
+			plugsDesktop = true
+			break
+		}
+	}
+	if !plugsDesktop {
+		return nil
+	}
+
+	// If there is no session bus, our job is done.  We check this
+	// manually to avoid dbus.SessionBus() auto-launching a new
+	// bus.
+	busAddress := osGetenv("DBUS_SESSION_BUS_ADDRESS")
+	if len(busAddress) == 0 {
+		return nil
+	}
+
+	conn, err := dbus.SessionBus()
+	if err != nil {
+		return err
+	}
+
+	portal := conn.Object("org.freedesktop.portal.Documents",
+		"/org/freedesktop/portal/documents")
+	var mountPoint []byte
+	if err := portal.Call("org.freedesktop.portal.Documents.GetMountPoint", 0).Store(&mountPoint); err != nil {
+		// It is not considered an error if
+		// xdg-document-portal is not available on the system.
+		if dbusErr, ok := err.(dbus.Error); ok && dbusErr.Name == "org.freedesktop.DBus.Error.ServiceUnknown" {
+			return nil
+		}
+		return err
+	}
```

Comment:
I do wonder if there is a cheap way to detect of the portal is running besides hitting DBus. I mean, this is correct and it isn't like application startup is all that zippy, but a cheap stat()-style test might be preferred and only if not running, hit the API to activate it (note that at this point, snap run is unconfined so it can probe freely). There might not be a robust way to do this but wanted to at least ask. Not a blocker

Reply:
This is the same method flatpak uses to ensure the document portal is running before trying to bind mount it:

https://github.com/flatpak/flatpak/blob/master/common/flatpak-run.c#L1811-L1862

Waking a running document portal process is probably not the end of the world, since it will likely be woken right afterwards when snap-confine/snap-update-ns set up the user mounts.